draft-ietf-ppm-dap.md

```diff
@@ -484,6 +480,138 @@ struct {
 } HpkeCiphertext;
 ~~~
 
+## Queries {#query}
+
+Aggregated results are computed based on sets of report, called batches. The
+Collector influences which reports are used in a batch via a "query." The
+Aggregators use this query to carry out the aggregation flow and produce
+aggregate shares encrypted to the Collector.
+
+This document defines the following query types:
+
+~~~
+enum {
+   reserved(0), // Reserved for testing purposes
+   time-interval(1),
+   fixed-size(2),
+   (65535)
+} QueryType;
+~~~
+
+The time-interval query type is described in {{time-interval-query}}; the
+fixed-size query type is described in {{fixed-size-query}}. Future
+specifications can introduce new query types as needed. A query includes
+parameters used by the Aggregators to select a batch of reports specific to the
+given query type. A query is defined as follows:
+
+~~~
+enum {
+    next-batch(0),
+    by-batch-id(1),
+} FixedSizeQueryType;
+
+opaque BatchId[32];
+
+struct {
+    FixedSizeQueryType fixed_size_query_type;
+    select (fixed_size_query_type) {
+        next-batch:  Empty;
```

Comment:
What is the idempotency story for `next-batch` requests? I think achieving idempotency requires the interaction to look something like this:

1. Collector issues `next-batch` collect request to Leader, receives collect job URI. Collector notes that the collect job URI needs to be collected to its own durable storage.
2. Collector polls collect job URI until the Leader gives a completed collect response, records the aggregation value & that the collect job is complete to its own durable storage.

If that's accurate, I think that we should spell out that `next-batch` returns the same collect job until the Collector actually collects the job--I think this isn't spell out currently, and if we don't have those semantics, repeated `next-batch` requests might cause batches to be dropped if recording the collect job URI to the Collector's durable storage fails.

Also, if it's acceptable to allow batch IDs to be enumerable (e.g. a counter, as suggested by others), I think a simpler idempotent design would be to have the Collector track which batch ID is next itself, and change `next-batch` to something like `max-batch` to allow Collectors to "start up" without knowing what the next batch is. This approach would simplify the Collector's state updates.

Reply:
> What is the idempotency story for `next-batch` requests? I think achieving idempotency requires the interaction to look something like this:
> 
>     1. Collector issues `next-batch` collect request to Leader, receives collect job URI. Collector notes that the collect job URI needs to be collected to its own durable storage.
> 
>     2. Collector polls collect job URI until the Leader gives a completed collect response, records the aggregation value & that the collect job is complete to its own durable storage.

This is accurate.

> If that's accurate, I think that we should spell out that `next-batch` returns the same collect job until the Collector actually collects the job--I think this isn't spelled out currently, and if we don't have those semantics, repeated `next-batch` requests might cause batches to be dropped if recording the collect job URI to the Collector's durable storage fails.

Each `next-batch` query would correspond to a different batch, and thus a different collect job. If the Collector queues up a bunch of collect jobs, then each job would still correspond to a different batch, right?


> Also, if it's acceptable to allow batch IDs to be enumerable (e.g. a counter, as suggested by others), I think a simpler idempotent design would be to have the Collector track which batch ID is next itself, and change `next-batch` to something like `max-batch` to allow Collectors to "start up" without knowing what the next batch is. This approach would simplify the Collector's state updates.

Reply:
(sorry, this comment is a little late to the party)

> Each next-batch query would correspond to a different

I see -- in that case, I think there is potential for data-loss if the Collector crashes at an inopportune time. Consider the following:

1. The Leader has a collect job ready to be retrieved, call it `X`.
2. The Collector issues a `next-batch` request, receives a collect job corresponding to `X`.
3. Before storing the collect job ID to durable storage or polling for completion, the Collector (or some relevant subcomponent) crashes & loses its state.

In this case, I think the Collector has lost batch `X` permanently: it no longer knows the relevant collect job ID, and there is no way to interact with the Leader to retrieve it since further `next-batch` requests will result in different collect jobs being returned.

I think the best way to fix this might be based on the idea of making the batch IDs enumerable as suggested by #301. I'll add a comment to that bug.